java/src/main/java/com/microsoft/onnxruntime/ONNXJavaType.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright © 2019, Oracle and/or its affiliates. All rights reserved.
+ */
+package com.microsoft.onnxruntime;
+
+import com.microsoft.onnxruntime.TensorInfo.ONNXTensorType;
+
+/**
+ * An enum over supported Java primitives (and String).
+ */
+public enum ONNXJavaType {
+    FLOAT(1, float.class, 4),
+    DOUBLE(2, double.class, 8),
+    INT8(3, byte.class, 1),
+    INT16(4, short.class, 2),
+    INT32(5, int.class, 4),
+    INT64(6, long.class, 8),
+    BOOL(7, boolean.class, 1),
+    STRING(8, String.class, 4),
+    UNKNOWN(0, Object.class, 0);
+
+    private static final ONNXJavaType[] values = new ONNXJavaType[9];
+
+    static {
+        for (ONNXJavaType ot : ONNXJavaType.values()) {
+            values[ot.value] = ot;
+        }
+    }
+
+    public final int value;
+    public final Class<?> clazz;
+    public final int size;
+
+    ONNXJavaType(int value, Class<?> clazz, int size) {
+        this.value = value;
+        this.clazz = clazz;
+        this.size = size;
+    }
+
+    /**
+     * Maps from an int in native land into an ONNXJavaType instance.
+     * @param value The value to lookup.
+     * @return The enum instance.
+     */
+    public static ONNXJavaType mapFromInt(int value) {
+        if ((value > 0) && (value < values.length)) {
+            return values[value];
+        } else {
+            return UNKNOWN;
+        }
+    }
+
+    /**
+     * Must match the values from ONNXUtil.c.
+     *
+     * @param onnxValue The value from ONNXUtil.c
+     * @return A JavaDataType instance representing the Java type
+     */
+    public static ONNXJavaType mapFromONNXTensorType(ONNXTensorType onnxValue) {
+        switch (onnxValue) {
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT8:
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT8:
+                return ONNXJavaType.INT8;
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT16:
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT16:
+                return ONNXJavaType.INT16;
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT32:
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32:
+                return ONNXJavaType.INT32;
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64:
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64:
+                return ONNXJavaType.INT64;
+            case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16:
```

Comment:
>ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT16 [](start = 17, length = 37)

Mapping this to FLOAT is probably wrong

Reply:
I think it upconverts an fp16 result into a float on the way out of the native code, but I'll double check tomorrow.

Reply:
Typically, this is represented as uint16_t in C++ code. Converting it to a standard float requires a call to an elaborate function.

---
In reply to: [340888816](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r340888816) [](ancestors = 340888816)

Reply:
Yes, in ONNXUtil.c it calls the convertHalfToFloat function whenever a fp16 value is passed back up. Given fp16 is usually used for memory and perf reasons within the network it doesn't seem so bad to have it be upconverted into a float for the output. At least that preserves all the information and doesn't require the user to do bit twiddling in Java land. If you want me to return it as a `short` I can do that too, but it'll be more confusing for the users.

Reply:
So what is end to end scenario? This allows to read output of float16, what about creating a Tensor of float16 as input? Also any plans to support BFloat16?

---
In reply to: [341115288](https://github.com/microsoft/onnxruntime/pull/2215#discussion_r341115288) [](ancestors = 341115288)

Reply:
Currently I don't have a way of creating a float16 Tensor as an input. That's mainly because I couldn't figure out a good carrier type in Java. I suppose I could put a flag on an entry point that takes a float array or buffer and have it generate either float or float16.

For bfloat16 I was going off the comments that said it wasn't supported in the onnx runtime. Has that changed?

Reply:
@Craigacp Since they're using `uint16_t` in C++ for those, we can simply use `short` in Java, that's perfectly fine. There's no fundamental type for that in C++ either! It works great, see examples of that here: https://github.com/bytedeco/javacpp/blob/master/src/test/java/org/bytedeco/javacpp/IndexerTest.java#L868